design/design-doc-aggregate-resources.md

```diff
@@ -0,0 +1,965 @@
+spec:
+  engineVersion: "5.7"
+  # The connection secret of the dynamically provisioned AggregateResource will
+  # be propagated to this reference.
+  writeConnectionSecretToRef:
+    name: sql
+```
+
+Creation of the above resource claim would trigger dynamic provisioning of the
+following resources:
+
+```yaml
+apiVersion: aggregation.crossplane.io/v1alpha1
+kind: AggregateResource
+metadata:
+  name: default-sql-gdm0w
+spec:
+  aggregationRule:
+    resourceSelector:
+      matchLabels:
+        # The dynamically provisioned AggregateResource matches all resources
+        # that were also dynamically provisioned by the resource claim that
+        # triggered its provisioning.
+        resourceclaim.crossplane.io/namespace: default
+        resourceclaim.crossplane.io/name: sql
+        resourceclaim.crossplane.io/uid: 2200b0c8-0da2-11ea-8d71-362b9e155667
+  # The AggregateResource maintains a claimRef back to the resource claim that
+  # triggered its dynamic provisioning.
+  claimRef:
+    apiVersion: database.crossplane.io/v1alpha1
+    kind: MySQLInstance
+    namespace: default
+    name: sql
+    uid: eabce854-0cd7-11ea-8d71-362b9e155667
+  # The AggregateResource maintains a classRef back to the aggregate resource
+  # class that was used to dynamically provision it.
+  classRef:
+    apiVersion: aggregation.crossplane.io/v1alpha1
+    kind: AggregateResourceClass
+    name: private-mysql-server
+    uid: ea414c74-0cd9-11ea-8d71-362b9e155667
+  # The aggregated connection secret will be written to this reference. In this
+  # case it will contain only the keys of the MySQLServer managed resource, as
+  # it is the only aggregated resource that publishes a connection secret.
+  writeConnectionSecretsToRef:
+    namespace: crossplane-system
+    name: ea414c74-0cd9-11ea-8d71-362b9e155667
+  reclaimPolicy: Delete
+status:
+  # Each object under the aggregatedResource represents a managed resource that
+  # matches the resourceSelector of this AggregateResource, including the
+  # status of the managed resource's binding to the AggregateResource.
+  aggregatedResources:
+  - apiVersion: azure.crossplane.io/v1alpha3
+    kind: ResourceGroup
+    name: default-sql-ab4k8
+    uid: bf44f148-0dd7-11ea-8d71-362b9e155667
+    bindingPhase: Bound
+  - apiVersion: database.azure.crossplane.io/v1beta1
+    kind: SQLServer
+    name: default-sql-d82nd
+    uid: e3c3166c-0dd7-11ea-8d71-362b9e155667
+    bindingPhase: Bound
+  - apiVersion: database.azure.crossplane.io/v1alpha3
+    kind: MySQLServerVirtualNetworkRule
+    name: default-sql-9dm3v
+    uid: e02cffd6-0dd7-11ea-8d71-362b9e155667
+    bindingPhase: Bound
+  # The AggregateResource's binding status represents its binding to the claim
+  # that triggered its dynamic provisioning. An AggregateResource binds to its
+  # claim immediately, then acts as a proxy for the claims binding to the
+  # managed resources it aggregates.
+  bindingPhase: Bound
```

Comment:
The idea is that the binding is transitive, and is "pushed down a level" when aggregate resources are involved. i.e. The most accurate way to know whether your `MySQLInstanceClaim` was bound to all the things it should be bound to is to inspect the `AggregateResource`, not the `MySQLInstanceClaim`.

There's two things to consider here:

* What does "bound" mean? Does it mean "I'm bound to this thing" or does it mean "I'm ready to be used". Today it's a blurred line. We wait for a managed resource to appear available before allowing a claim to bind to it, but if that managed resource becomes _unavailable_ somehow (e.g. due to an outage or maintainance) we don't _unbind_ it - that would not make sense. Instead it remains bound, but is not usable. My thinking is that the "bound" should probably mean "bound" and that readiness should be modelled separately.
* An `AggregateResource`aggregates using labels, so it doesn't have a fixed set of "all sub-resources"; that could change over time. It could say "all sub-resources that I currently match are ready", but then another one could come along and match later, and remain unready.

Reply:
>An AggregateResource aggregates using labels, so it doesn't have a fixed set of "all sub-resources"; that could change over time.

Oh I didn't get that from my reading. In fact, this sounds weird to me. Today if a resource is provisioned, even if its class changes that resource doesn't care because it's provisioned with the spec that was desired at the time that the creation was requested and that's solid design IMO. Why not keep a similar behavior for `AggregateResource`? If at the creation time I made the labels such that 5 resources are captured by that `AggregateResource`, I'd expect it to stay same unless I take specific action _on_ that `AggregateResource`. But as it seems, if another person creates a resource with those specific labels my `AggregateResource` will happily pick it up.

To be perfectly honest, I see too many moving pieces in general and so much care to be taken by the user. Maybe I'd get confidence when I use it, but right now I can't see myself using Aggregate mechanism in a prod/dev environment that is actively maintained by different people.

Reply:
I have a hard time seeing the issue here.

Labels as a way of grouping and selecting resources is a fundamental Kubernetes concept, that is widely used in the ecosystem, but whenever it comes up in Crossplane designs I feel like there's some degree of the FUD along the lines of "but what if our users don't use a good labelling scheme?". I've seen this on the KubernetesApplication, simple resource class selection, and aggregate resource designs.

`AggregateResource` labels could _in theory_ change over time - so we shouldn't design assuming they won't - but in reality it would be an edge case. A dynamically provisioned `AggregateResource` is labelled with it's resource claim's UID, so someone would have to actively add the resource claim's name, namespace, and UID as labels on another managed resource to get it to retroactively joint the `AggregateResource`. `ReplicaSet` works similarly - they match pods on labels but the labels are designed such that it's basically impossible that a `Pod` would _accidentally_ match the wrong `ReplicaSet`.

If folks decided to statically provision an `AggregateResource` the onus is on them to make sure the labels they use are likely to be unique. Perhaps they simulate the dynamic provisioning behaviour and aggregate using a UID.

Reply:
Sorry, I guess I just panicked when I realized it could change and I thought labels that `AggregateResource` uses are arbitrary like the one on `AggregateResourceClass`. But all three labels are generated by Crossplane from the claim it binds to, so, I see the case how it's hard for a user to mess this up. Taking it back 😅